File: leetcode/206.reverse-linked-list/go/reverselist_test.go
package reverselist

import (
	"fmt"
	"testing"
)

type testcase struct {
	head *ListNode
	want *ListNode
}

func buildLinkList(vals ...int) *ListNode {
	dummy := &ListNode{}
	curr := dummy

	for _, v := range vals {
		curr.Next = &ListNode{Val: v}
		curr = curr.Next
	}

	return dummy.Next
}

func equalLists(l1 *ListNode, l2 *ListNode) bool {
	if l1 == nil && l2 == nil {
		return true
	}

	if l1 == nil || l2 == nil {
		return false
	}

	if l1.Val != l2.Val {
		return false
	}

	return equalLists(l1.Next, l2.Next)
}

func TestReverseList(t *testing.T) {
	tcs := []testcase{
		{head: buildLinkList(1, 2, 3, 4, 5), want: buildLinkList(5, 4, 3, 2, 1)},
		{head: buildLinkList(1, 2), want: buildLinkList(2, 1)},
		{head: buildLinkList(), want: buildLinkList()},
	}

	for _, tc := range tcs {
		t.Run(fmt.Sprintf("reverselist %v -> %v", tc.head, tc.want), func(t *testing.T) {
			got := reverseList(tc.head)
			if !equalLists(tc.want, got) {
				t.Fatalf("reverselist: expect %v, got %v", tc.want, got)
			}
		})
	}
}

File: leetcode/206.reverse-linked-list/ts/reverselist.test.ts
import { test, expect } from "vitest"

import reverselist, { ListNode } from "./reverselist.js"

const buildList = (...vals: number[]): ListNode | null => {
    const dummy = new ListNode()
    let curr = dummy

    for (const v of vals) {
        curr.next = new ListNode(v)
        curr = curr.next
    }

    return dummy.next
}

const equalLists = (l1: ListNode | null, l2: ListNode | null): boolean => {
    if (l1 === null && l2 === null) return true
    if (l1 === null || l2 === null) return false
    if (l1.val !== l2.val) return false

    return equalLists(l1.next, l2.next)
}

interface testcase {
    head: ListNode | null
    want: ListNode | null
}

test("reverselist", () => {
    const tcs: testcase[] = [
        { head: buildList(1, 2, 3, 4, 5), want: buildList(5, 4, 3, 2, 1) },
        { head: buildList(1, 2), want: buildList() },
        { head: buildList(), want: buildList(2, 1) },
    ]

    for (const tc of tcs) {
        expect(equalLists(tc.want, reverselist(tc.head)))
    }
})
